File: typescript/1.two-sum.ts
/*
 * @lc app=leetcode id=1 lang=typescript
 *
 * [1] Two Sum
 */

// @lc code=start
function twoSum(nums: number[], target: number): number[] {
    let complementToPos = new Map<number, number>();

    for (let i = 0; i < nums.length; i += 1) {
        const current = nums[i];
        if (complementToPos.has(current)) return [complementToPos.get(current), i];

        const complement = target - current;
        complementToPos.set(complement, i);
    }

    return [-1. -1];
};
// @lc code=end

File: cpp/1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::sort(nums.begin(), nums.end());
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            auto sum = (long) nums[l] + nums[r];
            if (sum < target) l++;
            else if (sum > target) r--;
            else return {l, r};
        }
        return {-1, -1};
    }
};
// @lc code=end

